src/software/ai/hl/stp/play/offense/offense_play_test.py

```diff
@@ -67,18 +69,21 @@ def test_offense_play(simulated_test_runner):
     )
 
     # Always Validation
-    # TODO- #2779 Validation
-    always_validation_sequence_set = [[]]
+    inv_always_validation_sequence_set = [
+        [BallAlwaysStaysInRegion(regions=[field.fieldBoundary()])]
```

Comment:
this isn't a problem now, but in the future, I can see that we might accidentally kick the ball out and fail this test. But it's okay for now and we can move it to the ag_always validation set in the future if it's a problem

Reply:
My understanding based on the high-level goals of the test, as well as how its currently written: if the ball is kicked and misses the net, then the other validations will also fail since the possession is lost and the referee would reset the field since the ball missed the net.

It makes sense in the current state: invariant-always for ball possession & ball position within field, and aggregate-eventually for scoring the ball. 

Now what confuses me: what is the difference between always vs. eventually for invariant? by definition, an test invariant should always hold? first time I'm looking through the simulated test code so still figuring out the api...

Reply:
always: means that this function should always be true in the lifetime of the test
eventually: should be true at some point in the test

moving this validation to aggregate means that we get to track how often our offense_play kicks it out of bounds (failing this validation) without constantly failing CI